Laravel-Backpack/CRUD
# Allow other fields to use `subfields` attribute (not just `checklist_dependency`)

## WHY

Fixes https://github.com/Laravel-Backpack/CRUD/issues/4065 

### BEFORE - What was wrong? What was happening before this PR?

We couldn't use `subfields` for `repeatable` and `relationship` because they were used by `checklist_dependency` in a different way: 
- in `checklist_dependency`, the `subfields` were treated as regular fields by the CRUD; the two `subfields` were saved creating & updating, NOT the main field;
- in `repeatable` and `relationship` we want the _main field_ to be saved, not the subfields (actually, the subfields will be _part of_ the main field because they have the same name as it, just add stuff using bracket notation - eg. `testimonial[name]` and `category[body]`);

### AFTER - What is happening after this PR?

- `subfields` can be used the way `repeatable` and `relationship` want them (as fields with bracket names)
- `subfields` can STILL be used by `checklist_dependency`, because Backpack treats them as regular fields if the field NAME is an array;


## HOW

### How did you achieve that, in technical terms?

In short:
- if `$field['name']` is a `string`, `subfields` are considered inconsequential;
- if `field['name']` is an `array`, `subfields` are considered important, are checked for relationships;

This is probably not a long-term solution, but it's the only way I've found to mix this in a non-breaking way, without introducing a flag for `checklist_dependency` or anything.

It has the added benefit that what we coded for `date_range` helps in this respect. Only `date_range` and `checklist_dependency` use arrays as field names, and they both need the same functionality: save both inputs in the name array. In addition to the `date_range` field, what `checklist_dependency` needs extra is for the `subfields` to be checked for relationships, that's it.

### Is it a breaking change or non-breaking change?

Non-breaking. 

### How can we test the before & after?

Before - inside the Users CRUD check that roles & permissions work.
After - inside the Users CRUD check that roles & permissions work.

---

Sidenotes: 
- this does NOT make `repeatable` and `relationship` accept `subfields`; it just allows them to do so in a different PR;
- while working on this I discovered another bug in `checklist_dependency`; see #4071
